### ISSUE TITLE
eventlet closes fd under my feet

### ISSUE THREAD

**Issue by antocuni** [eventlet/eventlet#219]
The following code fails under PyPy (tested with the most recent 2.5.1) with an OSError

```
import eventlet
eventlet.monkey_patch()
import subprocess
import gc

p = subprocess.Popen(['ls'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
# the following line creates a _SocketDuckForFd(3) and .close()s it, but the
# object has not been collected by the GC yet
stdout, stderr = p.communicate()

f = open('/tmp/foo.txt', 'w')    # f.fileno() == 3
gc.collect() # this calls the __del__ of _SocketDuckForFd(3), close()ing it again

f.close() # OSError, because the fd 3 has already been closed
```

The problem is that `._close()` is called twice on a `greenio._SocketDuckForFd` object: the first close is done explicitly by `subprocess.communicate()`, while the second is done explicitly by the `__del__`.
Under CPython, the `__del__` is executed immediately; we STILL get an OSError because we close the same fd twice, but the exception is eaten by the try/except in `_SocketDuckForFd._close()`.
Under PyPy, the `__del__` is executed only later when the GC runs (in this case, the gc is triggered manually), but in the meantime the fd has been reused, and thus it closes a file which does not belong to us.

To fix the problem, it's enough to check whether we already closed:
```
    def _close(self):
        if self._closed:
            return
        notify_close(self._fileno)
        self._mark_as_closed()
        try:
            os.close(self._fileno)
        except:
            # os.close may fail if __init__ didn't complete
            # (i.e file dscriptor passed to popen was invalid
            pass
```

This is related to PyPy Issue 2005 (which we fixed today, and it's still present in PyPy 2.5.1):
https://bitbucket.org/pypy/pypy/issue/2005

The combined effect of the two problems is that if you call eventlet.monkey_patch() and subprocess.Popen during the importing, you might get the error described in the PyPy issue.

**Comment by temoto** [comment]
Hello, @antocuni thanks for bug report and patch. Please see if this version fixes the issue you described 2b6a7cef378ce64989d1531d9311890cd6c1cd6b
It actually contains a slightly modified version of your test code, but I'd like a second confirmation.

**Comment by antocuni** [comment]
Hello @temoto: yes, I confirm that the bug is fixed with your patch. Thank you!

**Comment by temoto** [comment]
FYI: fix is included in v0.17.2 release, available on PyPi.
